net/utils.cpp

```diff
@@ -304,6 +300,7 @@ class DefaultResolver : public Resolver {
 
     void discard_cache(const char *host, IPAddr ip) override {
         auto ipaddr = dnscache_.borrow(host);
+        scoped_rwlock _(*ipaddr, WLOCK);
         if (ip.undefined() || ipaddr->empty()) {
             ipaddr->delete_all();
```

Comment:
recycle will destruct ipaddrlist so do not need to delete items right here, and also do not need a write lock in such condition.